src/plugins/data/public/antlr/dql/code_completion.ts

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { CharStream, CommonTokenStream, TokenStream } from 'antlr4ng';
+import { CodeCompletionCore } from 'antlr4-c3';
+import { monaco } from '@osd/monaco';
+import { DQLLexer } from './.generated/DQLLexer';
+import { DQLParser, KeyValueExpressionContext } from './.generated/DQLParser';
+import { getTokenPosition } from '../shared/cursor';
+import { IndexPattern, IndexPatternField } from '../../index_patterns';
+import { getHttp } from '../../services';
+import { QuerySuggestionGetFnArgs } from '../../autocomplete';
+import { DQLParserVisitor } from './.generated/DQLParserVisitor';
+
+const findCursorIndex = (
+  tokenStream: TokenStream,
+  cursor: monaco.Position,
+  whitespaceToken: number
+): number | undefined => {
+  const cursorCol = cursor.column - 1;
+
+  for (let i = 0; i < tokenStream.size; i++) {
```

Comment:
I think it might mess with the readability. This function essentially just walks through the tokens in tokenStream and checks the start and end column to see if its on the current token, then returns that iterator variable w/ modifications. Using `find` might be good for exactly how this function is now, but finding this cursor index could be complicated if something within the grammar changes, so personally, I think its better to have something that clearly walks through the tokens instead of a callback function.